hy/scoping.py

```diff
@@ -0,0 +1,243 @@
+import ast
+import itertools
+
+from hy.lex import mangle
+from hy.models import Expression, List, Symbol
+
+import hy._compat
+
+
+class NodeWithName:
+    def __init__(self, node, index=None):
+        self.node = node
+        self.index = index
+
+    @property
+    def name(self):
+        if isinstance(self.node, (ast.Global, ast.Nonlocal)):
+            return self.node.names[self.index]
+        elif hy._compat.PY3_10:
+            if isinstance(self.node, (ast.MatchAs, ast.MatchStar)):
+                return self.node.name
+            elif isinstance(self.node, ast.MatchMapping):
+                return self.node.rest
+        return self.node.id
+
+    @name.setter
+    def name(self, new_name):
+        if isinstance(self.node, (ast.Global, ast.Nonlocal)):
+            self.node.names[self.index] = new_name
+        elif hy._compat.PY3_10:
+            if isinstance(self.node, (ast.MatchAs, ast.MatchStar)):
+                self.node.name = new_name
+            elif isinstance(self.node, ast.MatchMapping):
+                self.node.rest = new_name
+        self.node.id = new_name
+
+
+class Scope:
+    def __init__(self, stack, scope_id):
+        self.stack = stack
+        self.id = scope_id
+
+    def __enter__(self):
+        self.parent = self.stack.scope
+        self.stack.scope = self
+        return self
+
+    def __exit__(self, *args):
+        self.stack.scope = self.parent
+        return False
+
+    def is_global_scope(self):
+        return self.id == 0
+
+    def is_function_scope(self):
+        return isinstance(self, ScopeFn) and self.is_fn
+
+    def inside_function_scope(self):
+        cur = self
+        while cur.id != 0:
+            if cur.is_function_scope():
+                return True
+            cur = cur.parent
+        return False
+
+    def nearest_python_scope(self):
+        cur = self
+        while cur.id != 0 and isinstance(cur, ScopeLet):
+            cur = cur.parent
+        return cur
+
+    def access(self, node, index=None):
+        if not isinstance(node, NodeWithName):
+            node = NodeWithName(node, index)
+        self._access(node)
+
+    def assign(self, node, index=None):
+        if not isinstance(node, NodeWithName):
+            node = NodeWithName(node, index)
+        self._assign(node)
+
+    def _access(self, node):
+        pass
```

Comment:
is it valid to access these `pass` methods on an instance of `Scope` or should this raise an error?